engine/gamesys/src/gamesys/scripts/script_sys_gamesys.cpp

```diff
@@ -0,0 +1,476 @@
+// Copyright 2020-2024 The Defold Foundation
+// Copyright 2014-2020 King
+// Copyright 2009-2014 Ragnar Svensson, Christian Murray
+// Licensed under the Defold License version 1.0 (the "License"); you may not use
+// this file except in compliance with the License.
+//
+// You may obtain a copy of the License, together with FAQs at
+// https://www.defold.com/license
+//
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+#include <dlib/opaque_handle_container.h>
+
+#include <resource/resource.h>
+#include <resource/async/load_queue.h>
+
+#include "../gamesys.h"
+
+#include "script_buffer.h"
+#include "script_sys_gamesys.h"
+
+extern "C"
+{
+    #include <lua/lua.h>
+    #include <lua/lauxlib.h>
+}
+
+namespace dmGameSystem
+{
+    /*# System API documentation
+     *
+     * Functions and messages for using system resources, controlling the engine,
+     * error handling and debugging.
+     *
+     * @document
+     * @name System
+     * @namespace sys
+     */
+
+    enum RequestStatus
+    {
+        REQUEST_STATUS_ERROR_IO_ERROR  = -2,
+        REQUEST_STATUS_ERROR_NOT_FOUND = -1,
+        REQUEST_STATUS_INITALIZED      = 1,
+        REQUEST_STATUS_PENDING         = 2,
+        REQUEST_STATUS_FINISHED        = 3,
+    };
+
+    struct LuaRequest
+    {
+        dmScript::LuaCallbackInfo* m_CallbackInfo;
+        HOpaqueHandle              m_Handle;
+        dmBuffer::HBuffer          m_Payload;
+        dmLoadQueue::HRequest      m_Request;
+        const char*                m_Path;
+        dmhash_t                   m_PathHash;
+        RequestStatus              m_Status;
+    };
+
+    struct SysModule
+    {
+        dmResource::HFactory                m_Factory;
+        dmLoadQueue::HQueue                 m_LoadQueue;
+        dmOpaqueHandleContainer<LuaRequest> m_LoadRequests;
+        dmMutex::HMutex                     m_LoadRequestsMutex;
+    } g_SysModule;
+
+    // Assumes the g_SysModule.m_LoadRequestsMutex is held
+    static bool HandleRequestCompleted(LuaRequest* request)
+    {
+        // The request buffer can not be touched after this call
+        dmLoadQueue::FreeLoad(g_SysModule.m_LoadQueue, request->m_Request);
+        bool result = true;
+
+        if (dmScript::IsCallbackValid(request->m_CallbackInfo))
+        {
+            lua_State* L = dmScript::GetCallbackLuaContext(request->m_CallbackInfo);
+            DM_LUA_STACK_CHECK(L, 0);
+
+            // callback has the format:
+            // function(self, request_id, status, buffer)
```

Comment:
What do we want to return here? The request id can be used to identify a specific request that was issued, even if multiple requests for the same path was issued.

Reply:
Not sure what exactly is needed, but I suppose if we want to give ourselves a bit more flexibility then we could do what we do with `http.request()` and return a `result` object containing status, buffer and whatever else we may need.